ast/src/parsed/asm.rs

```diff
@@ -399,23 +400,111 @@ impl Machine {
                         MachineStatement::Pil(_, statement) => {
                             Box::new(statement.symbol_definition_names().map(|(s, _)| s))
                         }
-                        MachineStatement::CallSelectors(_, name) => Box::new(once(name)),
-                        MachineStatement::Degree(_, _)
-                        | MachineStatement::Submachine(_, _, _)
+                        MachineStatement::Submachine(_, _, _)
                         | MachineStatement::InstructionDeclaration(_, _, _)
                         | MachineStatement::LinkDeclaration(_, _)
                         | MachineStatement::FunctionDeclaration(_, _, _, _)
                         | MachineStatement::OperationDeclaration(_, _, _, _) => Box::new(empty()),
                     }
-                }),
+                })
+                .chain(self.arguments.names())
+                .chain(self.properties.names()),
         )
     }
 }
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Default, Clone)]
-pub struct MachineArguments {
+pub struct MachineArguments(pub Vec<Param>);
+
+impl MachineArguments {
+    pub fn names(&self) -> impl Iterator<Item = &String> {
+        self.0.iter().map(|p| &p.name)
+    }
+}
+
+impl TryFrom<Vec<Param>> for MachineArguments {
+    type Error = &'static str;
+
+    fn try_from(params: Vec<Param>) -> std::result::Result<Self, Self::Error> {
+        for p in &params {
+            if p.index.is_some() || p.ty.is_none() || p.name.is_empty() {
+                return Err("invalid machine argument");
+            }
+        }
+        Ok(MachineArguments(params))
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Default, Clone)]
+pub struct MachineProperties {
+    pub degree: Option<Expression>,
     pub latch: Option<String>,
     pub operation_id: Option<String>,
+    pub call_selectors: Option<String>,
+}
+
+impl MachineProperties {
+    pub fn names(&self) -> impl Iterator<Item = &String> {
+        self.call_selectors.iter()
+    }
+}
+
+impl TryFrom<Vec<(String, Expression)>> for MachineProperties {
+    type Error = &'static str;
+
+    fn try_from(prop_list: Vec<(String, Expression)>) -> std::result::Result<Self, Self::Error> {
+        let mut props: Self = Default::default();
+        for (name, value) in prop_list {
+            match name.as_str() {
+                "degree" => {
+                    if props.degree.is_some() {
+                        return Err("`degree` already defined");
+                    };
+                    props.degree = Some(value);
+                }
+                "latch" => {
```

Comment:
Can these three cases be combined?

Reply:
the issue is the Error needs to be a static string for bubbling up with lalrpop parsing AFAIR...
I thought about validating this in `machine_check` instead of parsing time, for more flexibility, will note this and try it in a follow up PR